doc/man3/str_to_bytes.3
.\"
.\" $Id: str_to_bytes.3,v 1.1 2000/07/27 16:59:03 alaffin Exp $
.\"
.\" Copyright (c) 2000 Silicon Graphics, Inc.  All Rights Reserved.
.\" 
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of version 2 of the GNU General Public License as
.\" published by the Free Software Foundation.
.\" 
.\" This program is distributed in the hope that it would be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.\" 
.\" Further, this software is distributed without any warranty that it is
.\" free of the rightful claim of any third person regarding infringement
.\" or the like.  Any license provided herein, whether implied or
.\" otherwise, applies only to this software file.  Patent licenses, if
.\" any, provided herein do not apply to combinations of this program with
.\" other software, or any other product whatsoever.
.\" 
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write the Free Software Foundation, Inc., 59
.\" Temple Place - Suite 330, Boston MA 02111-1307, USA.
.\" 
.\" Contact information: Silicon Graphics, Inc., 1600 Amphitheatre Pkwy,
.\" Mountain View, CA  94043, or:
.\" 
.\" http://www.sgi.com 
.\" 
.\" For further information regarding this notice, see: 
.\" 
.\" http://oss.sgi.com/projects/GenInfo/NoticeExplan/
.\"
.TH STR_TO_BYTES 3 07/25/2000 "Linux Test Project"
.SH NAME
str_to_bytes \- convert a string to a int byte count
str_to_lbytes \- convert a string to a long byte count
str_to_llbytes \- convert a string to a long long byte count
.SH SYNOPSIS
int str_to_byte(char *\fIstr\fR);
.br
long str_to_lbyte(char *\fIstr\fR);
.br
long long str_to_llbyte(char *\fIstr\fR);
.SH DESCRIPTION
\fBstr_to_bytes()\fR, \fBstr_to_lbytes()\fR, and \fBstr_to_llbytes()\fR converts
\fIstr\fR to an integer, long, or long long  byte count.  \fIstr\fR is an
floating point number optionally followed by a single character multiplier.
Currently the following multipliers are supported:
.sp
.nf
    Char  Meaning       Multiplier
    ----  ---------     --------------------------------
    b     Blocks        BSIZE or UBSIZE
    k     Kilobytes     2^10 (1024)
    K     Kilowords     2^10 (1024) * sizeof(long)
    m     Megabytes     2^20 (1048576)
    M     Megawords     2^20 (1048576) * sizeof(long)
    g     Gigabytes     2^30 (1073741824)
    G     Gigawords     2^30 (1073741824) * sizeof(long)
.fi
.sp
\fIstr\fR is interpreted as floating point number (base 10).
When using \fBstr_to_llbytes()\fR, the uppercase suffix will result
in multiplying by the size of a (long long) or 8.
.SH RETURNS
-1 if the integer portion of \fIstr\fR is invalid, if an unsupported
multiplier is supplied, or if \fIstr\fR has extra leading or trailing
characters.  If \fIstr\fR contains a negative number, the return
value will be negative.
.SH EXAMPLES
\fBstr_to_bytes("1000")\fR
.br
.RS 8
Returns 1000
.RE
.br
\fBstr_to_bytes("5b")\fR
.br 
.RS 8 
Returns 5 * BSIZE.
.RE
\fBstr_to_bytes("1.5m")\fR
.br
.RS 8
Returns 1.5 * 1048576 or 1572864
.RE

.SH LIMITATIONS

\fBstr_to_bytes()\fR and \fBstr_to_lbytes()\fR when compiled as
a 32 bit IRIX binary can only return a max number of 2g (2147483647).
However, \fBstr_to_lbytes()\fR is not limited by the 2g limit when
compiled as 64 bit binary, where \fBstr_to_bytes()\fR still is limited.

Note that the size of long will vary depending how if compiled as
a 32 or 64 bit binary.  The size of a long long is always 8.

Also note that on a traditional CRAY system, a block is 4096 bytes, where
on most IRIX systems a block is 1024 bytes.



ChangeLog
LTP-20020307
------------
o	fixed IDcheck.sh to prompt for IDs only when missing (Jay Huie)
o	added time duration option to pan (Jay Huie)
o	fixed bug [ #516577 ] ftruncate03 test (Jay Huie)
o	added 4 more cases to getcwd02 test (Jay Huie)
o	added time-schedule tool by Richard Gooch under sched test
	section (Manoj Iyer)
o	added trace_sched tool under sched test section (Manoj Iyer)
o	fixed bug [ #511427 ] pread02 test (Paul Larson)
o	fixed bug [ #523055 ] sched_getscheduler test (Paul Larson)
o	fixed bug [ #523137 ] sched_setscheduler02 test (Paul Larson)
o	fixed bug [ #525688 ] sendfile01 (Paul Larson)
o	added sigpending02 test to syscalls section (Paul Larson)
o       created a KNOWN-FAILURES file to document known 
        test failures (Robbie Williamson)
o	added sem02 test to ipc section (Robbie Williamson)
o	added SCTP section to network tests (Robbie Williamson)
o	fixed bug [ #505515 ] perf_lan6 test (Robbie Williamson)
o	added disktest by Brent Yardley to io section (Robbie Williamson)
o	closed the following bugs:
	491285, 491286, 506536, 514408 (Paul Larson and Robbie Williamson)


LTP-20020207
------------
o       added support for cross-compiling (Todd Inglett)
o       added LKML's cache_leak testcase to ltp/scratch (Nate Straz)
o       added IPv6 support (Robbie Williamson)
o       added "gethost" to /tools (Robbie Williamson)
o	fixed the race conditions in the float tests and removed 
	the sleeps (Robbie Williamson)
o       enabled non-root make authority (Paul Larson)
o       separated compilation into "make" and "make install" (Paul Larson)
o       added ipc_stress test (Manoj Iyer)
o       added pthreads_stress test (Manoj Iyer)
o       made changes to support architecture independence (Manoj Iyer & Paul Larson)
o	closed the following bugs: 
	504960, 505108, 504613, 504616, 491283, 506689, 508055, 506692, 508074
	491289, 506662, 511383, 511391, 511427, 511494, 504649, 514050, 512953
	(Manoj Iyer, Paul Larson, and Robbie Williamson)	


LTP-20020108
------------
o	fixed IDcheck.sh to work with bash1 (Nate Straz)
o	updated menu script (Robbie Williamson)
o	reset errno to 0 at the beginning of TEST macro (Paul Larson)
o	added floating point stress tests (Bull/IBM)
o	added fsx-linux NFS tests (Apple)	
o	define GLIBC_SIGACTION_BUG for sigaction02 (Robbie Williamson)
o	removed bogus testcase from dup201 test (Paul Larson)
o	fixed recvfrom01 case 5 (Wayne Boyer)
o	workaround for smp issue with waitpid12 (Manfred Spraul)


LTP-20011206
------------
o	necessary users/groups can now be created on installation with 
	user's permission
o	added a simple menu-based interface for running the LTP testsuite
o	fixed negative duration in pan output when -l isn't used
o	new set of tests under fs-maim
o	fixed race condition in nfslock01
o	ar01, ld01, ldd01, nm01, objdump01, and size01 fixed when multiple 
	copies are run simultaneously
o	workaround for SIGTTOU hang in ioctl02
o	shmget03 uses IPC_PRIVATE to make it safe when running multiple copies
o	compiler warnings and other minor errors fixed in many tests

LTP-20011107
------------
o	Many improvements to mtest05 and mtest06 tests
o	new test nfsstress
o	included ver_linux in LTP and made it run at the end of test scripts
o	check for necessary users/groups in Makefile and warn if they don't
	exist
o	documented the users and groups necessary for the testcases to run
o	simplified telnet01 when looking for root prompt
o	removed incorrect testcase from sendfile03
o	fixed modify_ldt01 test problems on newer kernels
o	added setrlimit to unlimit core size in waitpid05 test to work 
	around systems where this is set to 0
o	fixed mmstress pthread hang problem

LTP-20010925
------------
o	Testcases have been organized into categories in a directory 
	structure under the testcases subdirectory
o	About 400 new testcases added covering various system calls
o	Andreas Jaeger got confirmation from the glibc people that the
	ulimit test was broken as a result of some changes made in 
	glibc2.2.5.  Since it's the test that's broken, Nate Straz 
	removed the offending testcase.
o	Updated README and INSTALL files
o	Descriptions of the testcases have been added to the doc 
	directory and are organized by type (syscalls.txt, fs.txt, etc.)

------------
2000-09-18 Juan J. Quintela <quintela@fi.udc.es>
           Aaron Laffin <alaffin@sgi.com>

  * tests/mmap001.c: first checkin of a modified version of
  mmap001.c from the memtest suite.

2000-09-08  Egor Duda <deo@logos-m.ru>

	* tests/fpathconf01.c: Close opened file at exit.
	* tests/select01.c: Ditto.

2000-09-08  Andrea Arcangeli <andrea@suse.de>
  * tests/fcntl09.c, tests/fcntl10.c: Rearrange the F_SETLK test to
  to correctly use F_RDLCK and F_WRLCK.

2000-09-06  Egor Duda  <deo@logos-m.ru>

	* lib/parse_opts.c (parse_opts): Initialize allocated string
	to prevent heap corruption.



lib/str_to_bytes.c
/*
 * Copyright (c) 2000 Silicon Graphics, Inc.  All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it would be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 *
 * Further, this software is distributed without any warranty that it is
 * free of the rightful claim of any third person regarding infringement
 * or the like.  Any license provided herein, whether implied or
 * otherwise, applies only to this software file.  Patent licenses, if
 * any, provided herein do not apply to combinations of this program with
 * other software, or any other product whatsoever.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write the Free Software Foundation, Inc., 59
 * Temple Place - Suite 330, Boston MA 02111-1307, USA.
 *
 * Contact information: Silicon Graphics, Inc., 1600 Amphitheatre Pkwy,
 * Mountain View, CA  94043, or:
 *
 * http://www.sgi.com
 *
 * For further information regarding this notice, see:
 *
 * http://oss.sgi.com/projects/GenInfo/NoticeExplan/
 */
#include <stdio.h>
#include <sys/param.h>
#include "str_to_bytes.h"

/****************************************************************************
 * str_to_bytes(s)
 *
 * Computes the number of bytes described by string s.  s is assumed to be
 * a base 10 positive (ie. >= 0) number followed by an optional single
 * character multiplier.  The following multipliers are supported:
 *
 *              char    mult
 *              -----------------
 *              b       BSIZE  or BBSIZE
 *              k       1024 bytes
 *              K       1024 * sizeof(long)
 *              m       2^20 (1048576)
 *              M       2^20 (1048576 * sizeof(long)
 *              g       2^30 (1073741824)
 *              G       2^30 (1073741824) * sizeof(long)
 *
 * for instance, "1k" and "1024" would both cause str_to_bytes to return 1024.
 *
 * Returns -1 if mult is an invalid character, or if the integer portion of
 * s is not a positive integer.
 *
 ****************************************************************************/

#if CRAY
#define B_MULT	BSIZE		/* block size */
#elif sgi
#define B_MULT	BBSIZE		/* block size */
#elif defined(__linux__) || defined(__sun) || defined(__hpux)
#define B_MULT	DEV_BSIZE	/* block size */
#elif defined(_AIX)
#define B_MULT UBSIZE
#endif


#define K_MULT	1024		/* Kilo or 2^10 */
#define M_MULT	1048576		/* Mega or 2^20 */
#define G_MULT	1073741824	/* Giga or 2^30 */
#define T_MULT	1099511627776	/* tera or 2^40 */

int
str_to_bytes(s)
char    *s;
{
    char    mult, junk;
    int	    nconv;
    float   num;

    nconv = sscanf(s, "%f%c%c", &num, &mult, &junk);
    if (nconv == 0 || nconv == 3)
	return -1;

    if (nconv == 1)
	return num;

    switch (mult) {
    case 'b':
		return (int)(num * (float)B_MULT);
    case 'k':
		return (int)(num * (float)K_MULT);
    case 'K':
		return (int)((num * (float)K_MULT) * sizeof(long));
    case 'm':
		return (int)(num * (float)M_MULT);
    case 'M':
		return (int)((num * (float)M_MULT) * sizeof(long));
    case 'g':
		return (int)(num * (float)G_MULT);
    case 'G':
    		return (int)((num * (float)G_MULT) * sizeof(long));
    default:
	return -1;
    }
}

long
str_to_lbytes(s)
char    *s;
{
    char    mult, junk;
    long    nconv;
    float   num;

    nconv = sscanf(s, "%f%c%c", &num, &mult, &junk);
    if (nconv == 0 || nconv == 3)
	return -1;

    if (nconv == 1)
	return (long)num;

    switch (mult) {
    case 'b':
		return (long)(num * (float)B_MULT);
    case 'k':
		return (long)(num * (float)K_MULT);
    case 'K':
		return (long)((num * (float)K_MULT) * sizeof(long));
    case 'm':
		return (long)(num * (float)M_MULT);
    case 'M':
		return (long)((num * (float)M_MULT) * sizeof(long));
    case 'g':
		return (long)(num * (float)G_MULT);
    case 'G':
    		return (long)((num * (float)G_MULT) * sizeof(long));
    default:
	return -1;
    }
}

/*
 * Force 64 bits number when compiled as 32 IRIX binary.
 * This allows for a number bigger than 2G.
 */

long long
str_to_llbytes(s)
char    *s;
{
    char    mult, junk;
    long    nconv;
    double  num;

    nconv = sscanf(s, "%lf%c%c", &num, &mult, &junk);
    if (nconv == 0 || nconv == 3)
	return -1;

    if (nconv == 1)
	return (long long)num;

    switch (mult) {
    case 'b':
		return (long long)(num * (float)B_MULT);
    case 'k':
		return (long long)(num * (float)K_MULT);
    case 'K':
		return (long long)((num * (float)K_MULT) * sizeof(long long));
    case 'm':
		return (long long)(num * (float)M_MULT);
    case 'M':
		return (long long)((num * (float)M_MULT) * sizeof(long long));
    case 'g':
		return (long long)(num * (float)G_MULT);
    case 'G':
    		return (long long)((num * (float)G_MULT) * sizeof(long long));
    default:
	return -1;
    }
}

#ifdef UNIT_TEST

main(int argc, char **argv)
{
    int ind;

    if (argc == 1) {
	fprintf(stderr, "missing str_to_bytes() parameteres\n");
	exit(1);
    }

    for (ind=1; ind<argc; ind++) {

	printf("str_to_bytes(%s) returned %d\n",
	    argv[ind], str_to_bytes(argv[ind]));

	printf("str_to_lbytes(%s) returned %ld\n",
	    argv[ind], str_to_lbytes(argv[ind]));

	printf("str_to_llbytes(%s) returned %lld\n",
	    argv[ind], str_to_llbytes(argv[ind]));
    }
}

#endif
